### DISCUSSION TITLE
Associated Type

### DISCUSSION THREAD

**Discussion by AlseinX** [dotnet/csharplang#5222] [Language Ideas]
## Summary

As long as static abstract member in interfaces (#4436) comes to preview, interface is not only an abstraction over objects, but also types.
Associated type is also useful for a expressive type system. The associated type is an open generic type that is declared in an interface but specified by the implementation type.

eg:

Here is a simple definition of `IEnumerable<T>`

```
interface IEnumerable<out T>
{
    IEnumerator<T> GetEnumerator();
}
```

`GetEnumerator` returns an enumerator of an interface type, which means:

+ If the implementation is a value type, it must be boxed to be returned as an interface.

+ Members of the enumerator are virtual calls, which means performance cost and/or optimization complexity.

Associated type could make the enumerator a known type as long as the enumerable has a known type during compilation.

## Specification

Here is an probable demo of associated type, borrowing the syntax form rust:

```
interface IEnumerable<out T>
{
    type Enumerator: IEnumerator<T>;
    Enumerator GetEnumerator();
}
```

`type Enumerator` declares a type that must be specified by the implementation, and `: IEnumerator<T>` constraints it to be a subtype of `IEnumerator<T>`.

The use of the associated type must be at output position (the same rule as `out` covariant generic type)

And the implementation could be like this:

```
record MyType<T>(T A, T B) : IEnumerable<T>
{
    type IEnumerable<T>.Enumerator = Enumerator;
    public Enumerator GetEnumerator() => new(this);
    public struct Enumerator : IEnumerator<T>
    {
        private readonly MyType<T> _target;
        private int _position = -1;
        public Enumerator(MyType<T> target) => _target = target;
        public bool MoveNext() => ++_position > 1;
        public T Current =>
            _position == 0 ? _target.A :
            _position == 1 ? _target.B :
            throw new InvalidOperationException();
        public void Dispose() { }
    }
}
```

Then we could use it like:

```
void PrintValues<T, TEnumerable>(TEnumerable source)
    where TEnumerable: IEnumerable<T>
{
    using var enumerator = source.GetEnumerator();  // The type is resolved to the associated type `TEnumerable.Enumerator`
    while(enumerator.MoveNext())
    {
        Console.WriteLine(enumerator.Current);
    }
}
```

## Questions To Discuss:

+ What would the associated type be resolved to, when the interface is used as a type, rather than a generic constraint?

For backward compatibility, the constraint of an associated type might be limited to a unique interface or base class, which could be used as the resolved type, just like how we did before, but if there is no constraint, the resolved type defaults to `object`:

```
void PrintValues<T>(IEnumerable<T> source)
{
    using var enumerator = source.GetEnumerator();  // The type is resolved to IEnumerable<T> 
    while(enumerator.MoveNext())
    {
        Console.WriteLine(enumerator.Current);
    }
}
```

Another choice is to allow multiple interface constraints but one default type must be specified, which has to be one of these constraint super types or their common super types, UNLESS intersection types(#399) are supported.

+ Omitting associated type in implementations?

It might be appropriate to allow inferring the associated type by the return types of the implementation, so that changing the return types from a raw interface type into a associated type constrained by the interface would not break the existing implementation types.

Another way is that we could make it possible to allow the nested type `Enumerator` implicitly implements the associated type matching by the names, as we can implicitly implement methods and properties, so we won't need `type IEnumerable<T>.Enumerator = Enumerator;`

+ Input position usage of associated types?

Associated type is logically an output of a type during compilation, so I don't think it necessary to support it, otherwise it could make it impossible to use the interface as a type, because there might not be an appropriate type of an input parameter if the implementation type is unknown during compilation.

## Alternative Way

We could make no change to interface definitions and simply allow implementations to implement the members (methods that have interface return types or get-only properties of interface types) with given subtypes. And when using an interface as a generic constraint simply resolves the output value to the given subtype by the implementation, so that the returned values are not boxed or accessed via virtual table. The pro is that this does not add any new syntax or keywords.

**Comment by 333fred** [comment]
Duplicate of https://github.com/dotnet/csharplang/issues/1328.

**Reply by AlseinX** [reply]
> Duplicate of #1328.

I see, but I do think this post valuable for the suggested details.
